### PR TITLE
Finalizes Duplex APIs for gRPC

### PR DESCRIPTION
**Pull request by oldergod**

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/OkHttpClient.java [hunk 1]
```diff
@@ -197,21 +196,13 @@ public void apply(ConnectionSpec tlsConfiguration, SSLSocket sslSocket, boolean
...
-        responseBuilder.sinkAndCodec(sink, httpCodec);
-      }
-
-      @Override public BufferedSink sink(Response response) {
-        return response.sink();
-      }
-
-      @Override public boolean isDuplex(Request request) {
-        return request.isDuplex();
+      @Override public void duplex(Request.Builder requestBuilder, String method,
```
[the remark this answers is not in the document]
Comment (oldergod) [reply okhttp/src/main/java/okhttp3/OkHttpClient.java]: It's not used anyway at the moment,
we have the following in `Request.Builder` 
```java
    Builder duplex(String method, @Nullable MediaType contentType, long pipeMaxBufferSize) {
      return method(method, new DuplexRequestBody(contentType, pipeMaxBufferSize));
    }
```
but in tests we create the body ourselves as
```java
 .method("POST", new DuplexRequestBody(null, 1024L))
```
If the caller uses `DuplexRequestBuilder` anyway, maybe we can remove the duplex method inside the builder as well?

File: okhttp/src/main/java/okhttp3/internal/duplex/DuplexRequestBody.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    notifyAll();
+  }
+
+  @Override public synchronized void onResponse(Call call, Response response) {
+    if (this.failure != null || this.response != null) throw new IllegalStateException();
+    this.response = response;
+    notifyAll();
+  }
+
+  public synchronized Response awaitExecute() throws IOException {
```
[the remark this answers is not in the document]
Comment (oldergod) [reply okhttp/src/main/java/okhttp3/internal/duplex/DuplexRequestBody.java]: How about calling `createSink()` multiple times? Can the body be used for different `Call`s ?

File: okhttp/src/main/java/okhttp3/OkHttpClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (oldergod) [reply okhttp/src/main/java/okhttp3/OkHttpClient.java]: removed it for now

File: okhttp/src/main/java/okhttp3/internal/duplex/DuplexRequestBody.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import okhttp3.Callback;
+import okhttp3.MediaType;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okio.BufferedSink;
+import okio.Okio;
+import okio.Pipe;
+import okio.Sink;
+
+public final class DuplexRequestBody extends RequestBody implements Callback {
```
[the remark this answers is not in the document]
Comment (oldergod) [reply okhttp/src/main/java/okhttp3/internal/duplex/DuplexRequestBody.java]: wrote that down in my private diary

[the remark this answers is not in the document]
Comment (digitalbuddha) [reply okhttp/src/main/java/okhttp3/internal/duplex/DuplexRequestBody.java]: Is there anything need pertaining to interactions with the dispatcher? Should there be an ability to configure a different queue for duplex requests?

[the remark this answers is not in the document]
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/duplex/DuplexRequestBody.java]: Good call. Probably worth some docs at a minimum.

File: okhttp/src/main/java/okhttp3/internal/duplex/DuplexRequestBody.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/duplex/DuplexRequestBody.java]: As designed, it's one time use only. I don't think the utility of making multiple calls with it is worth the complexity that'd bring.

**Comment by swankjesse** [comment]
Yay!

**Pull request merged**